packages/authentication/src/__tests__/acceptance/jwt-auth-extension.acceptance.ts

```diff
@@ -0,0 +1,406 @@
+// Copyright IBM Corp. 2019. All Rights Reserved.
+// Node module: @loopback/authentication
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {inject} from '@loopback/context';
+import {addExtension, Application} from '@loopback/core';
+import {get} from '@loopback/openapi-v3';
+import {RestServer} from '@loopback/rest';
+import {Client, createClientForHandler, expect} from '@loopback/testlab';
+import {authenticate, AuthenticationBindings, UserProfile} from '../..';
+import {getApp, getUserRepository} from '../fixtures/helper';
+import {JWTAuthenticationStrategyBindings, USER_REPO} from '../fixtures/keys';
+import {MyAuthenticationSequence} from '../fixtures/sequences/authentication.sequence';
+import {JWTService} from '../fixtures/services/jwt-service';
+import {JWTAuthenticationStrategy} from '../fixtures/strategies/jwt-strategy';
+import {UserRepository} from '../fixtures/users/user.repository';
+
+describe('JWT Authentication', () => {
+  let app: Application;
+  let server: RestServer;
+  let test_users: UserRepository;
+
+  beforeEach(givenAServer);
+  beforeEach(givenAuthenticatedSequence);
+  beforeEach(givenProviders);
+
+  it('authenticates successfully with valid token', async () => {
+    class InfoController {
+      constructor(
+        @inject(JWTAuthenticationStrategyBindings.TOKEN_SERVICE)
+        public tokenService: JWTService,
+        @inject(USER_REPO)
+        public users: UserRepository,
+      ) {}
+
+      @get('/login')
+      async logIn() {
+        //
+        // ...Other code for verifying a valid user (e.g. basic or local strategy)...
+        //
+
+        // Now with a valid userProfile, let's create a JSON web token
+        const joeUser = this.users.list['joe@example.com'].user;
+
+        const joeUserProfile = {
+          id: joeUser.id,
+          email: joeUser.email,
+          name: `${joeUser.firstname} ${joeUser.surname}`,
+        };
+
+        return await this.tokenService.generateToken(joeUserProfile);
+      }
+
+      @get('/whoAmI')
+      @authenticate('jwt')
+      whoAmI(
+        @inject(AuthenticationBindings.CURRENT_USER) userProfile: UserProfile,
+      ) {
+        if (userProfile) {
```

Comment:
Suggestion (for here and any other instances of this code):
```ts
if(!userProfile) return 'userProfile is undefined';
if(!userProfile.email) return 'userProfile email is undefined';
return userProfile.email;
```

Reply:
I personally don't like this way of writing if statements.

Reply:
What's suggested by @nabdelgadir is consistent with our coding style. The key is to return fast to avoid deep if/else nesting/indentation. It's for readability.